Mathlib/GroupTheory/GroupAction/Hom.lean

```diff
@@ -1,212 +1,524 @@
+* `DistribMulActionHom φ A B`,
+  the type of equivariant additive monoid homomorphisms from `A` to `B`,
+  where `φ : M → N` is a morphism of monoids,
+  `M` acting on the additive monoid `A` and `N` acting on the additive monoid of `B`
+* `SMulSemiringHom φ R S`, the type of equivariant ring homomorphisms
+  from `R` to `S`, where `φ : M → N` is a morphism of monoids,
+  `M` acting on the ring `R` and `N` acting on the ring `S`.
 
 The above types have corresponding classes:
-* `SMulHomClass F M X Y` states that `F` is a type of bundled `X → Y` homs
-  preserving scalar multiplication by `M`
-* `DistribMulActionHomClass F M A B` states that `F` is a type of bundled `A → B` homs
-  preserving the additive monoid structure and scalar multiplication by `M`
-* `MulSemiringActionHomClass F M R S` states that `F` is a type of bundled `R → S` homs
-  preserving the ring structure and scalar multiplication by `M`
+* `MulActionHomClass F φ X Y` states that `F` is a type of bundled `X → Y` homs
+  which are `φ`-equivariant
+* `DistribMulActionHomClass F φ A B` states that `F` is a type of bundled `A → B` homs
+  preserving the additive monoid structure and `φ`-equivariant
+* `SMulSemiringHomClass F φ R S` states that `F` is a type of bundled `R → S` homs
+  preserving the ring structure and `φ`-equivariant
 
 ## Notations
 
-* `X →[M] Y` is `MulActionHom M X Y`.
-* `A →+[M] B` is `DistribMulActionHom M A B`.
-* `R →+*[M] S` is `MulSemiringActionHom M R S`.
+* `X →ₑ[φ] Y` is `MulActionHom φ X Y`.
+* `A →ₑ+[φ] B` is `DistribMulActionHom φ A B`.
+* `R →ₑ+*[φ] S` is `MulSemiringActionHom φ R S`.
+
+* When `M = N` and `φ = MonoidHom.id M`, we provide the notation `X →[M] Y`
+-/
+
+section CompTriple
+
+/-- Class of composing triples -/
+class CompTriple  {M N P : Type _} (φ : M → N) (ψ : N → P) (χ : outParam (M → P)) : Prop where
+  /-- The maps form a commuting triangle -/
+  comp_eq : ψ.comp φ = χ
+
+attribute [simp] CompTriple.comp_eq
+
+namespace CompTriple
+
+/-- Class of Id maps -/
+class isId {M : Type*} (σ : M → M) : Prop where
+  eq_id : σ = id
+
+instance {M : Type*} : isId (@id M) where
+  eq_id := rfl
+
+lemma comp_id {N P : Type _} {φ : N → N} [isId φ] {ψ : N → P} :
+    CompTriple φ ψ ψ := {
+  comp_eq := by simp only [isId.eq_id, Function.comp.right_id]}
+
+lemma id_comp {M N : Type _} {φ : M → N} {ψ : N → N} [isId ψ] :
+    CompTriple φ ψ φ := {
+  comp_eq := by simp only [isId.eq_id, Function.comp.left_id]}
+
+lemma comp {M N P : Type _}
+    {φ : M → N} {ψ : N → P} :
+  CompTriple φ ψ  (ψ.comp φ) := {comp_eq := rfl}
+
+lemma comp_inv {M N : Type _} {φ : M → N} {ψ : N → M}
+    (h : Function.RightInverse φ ψ) {χ : M → M} [isId χ] :
+    CompTriple φ ψ χ := {
+  comp_eq := by simp only [isId.eq_id, h.id] }
+
+lemma apply {M N P : Type _}
+    {φ : M → N} {ψ : N → P} {χ : M → P} (h : CompTriple φ ψ χ) (x : M) :
+  ψ (φ x) = χ x := by
+  rw [← h.comp_eq, Function.comp_apply]
+
+end CompTriple
+
+section MonoidHomCompTriple
```

Comment:
Similarly this section can go in `Mathlib/Algebra/Group/CompTypeclasses.lean`.

Reply:
CompTypeclass (plural?)